cpp/src/parquet/statistics.cc

```diff
@@ -32,84 +32,362 @@ using arrow::MemoryPool;
-      max_buffer_(AllocateBuffer(pool_, 0)) {
-  IncrementNumValues(num_values);
-  IncrementNullCount(null_count);
-  IncrementDistinctCount(distinct_count);
-
-  Copy(min, &min_, min_buffer_.get());
-  Copy(max, &max_, max_buffer_.get());
-  has_min_max_ = true;
-}
+template <typename DType, bool is_signed>
+struct CompareHelper {
+  typedef typename DType::c_type T;
+  static inline bool Compare(int type_length, const T& a, const T& b) { return a < b; }
+};
 
-template <typename DType>
-TypedRowGroupStatistics<DType>::TypedRowGroupStatistics(
-    const ColumnDescriptor* schema, const std::string& encoded_min,
-    const std::string& encoded_max, int64_t num_values, int64_t null_count,
-    int64_t distinct_count, bool has_min_max, MemoryPool* pool)
-    : pool_(pool),
-      min_buffer_(AllocateBuffer(pool_, 0)),
-      max_buffer_(AllocateBuffer(pool_, 0)) {
-  IncrementNumValues(num_values);
-  IncrementNullCount(null_count);
-  IncrementDistinctCount(distinct_count);
-
-  SetDescr(schema);
-
-  if (!encoded_min.empty()) {
-    PlainDecode(encoded_min, &min_);
+template <>
+struct CompareHelper<Int96Type, true> {
+  static inline bool Compare(int type_length, const Int96& a, const Int96& b) {
+    // Only the MSB bit is by Signed comparison
+    // For little-endian, this is the last bit of Int96 type
+    const int32_t amsb = static_cast<const int32_t>(a.value[2]);
+    const int32_t bmsb = static_cast<const int32_t>(b.value[2]);
+    if (amsb != bmsb) {
+      return (amsb < bmsb);
+    } else if (a.value[1] != b.value[1]) {
+      return (a.value[1] < b.value[1]);
+    }
+    return (a.value[0] < b.value[0]);
   }
-  if (!encoded_max.empty()) {
-    PlainDecode(encoded_max, &max_);
+};
+
+template <>
+struct CompareHelper<ByteArrayType, true> {
+  static inline bool Compare(int type_length, const ByteArray& a, const ByteArray& b) {
+    const int8_t* aptr = reinterpret_cast<const int8_t*>(a.ptr);
+    const int8_t* bptr = reinterpret_cast<const int8_t*>(b.ptr);
+    return std::lexicographical_compare(aptr, aptr + a.len, bptr, bptr + b.len);
   }
-  has_min_max_ = has_min_max;
-}
+};
 
-template <typename DType>
-bool TypedRowGroupStatistics<DType>::HasMinMax() const {
-  return has_min_max_;
-}
+template <>
+struct CompareHelper<FLBAType, true> {
+  static inline bool Compare(int type_length, const FLBA& a, const FLBA& b) {
+    const int8_t* aptr = reinterpret_cast<const int8_t*>(a.ptr);
+    const int8_t* bptr = reinterpret_cast<const int8_t*>(b.ptr);
+    return std::lexicographical_compare(aptr, aptr + type_length, bptr,
+                                        bptr + type_length);
+  }
+};
 
-template <typename DType>
-void TypedRowGroupStatistics<DType>::SetComparator() {
-  comparator_ =
-      std::static_pointer_cast<CompareDefault<DType> >(Comparator::Make(descr_));
+template <>
+struct CompareHelper<Int32Type, false> {
+  static inline bool Compare(int type_length, int32_t a, int32_t b) {
+    const uint32_t ua = a;
```

Comment:
Are you sure this is what you want? How is this handled on the writer side?

Reply:
This is replicated from the original location, if it's wrong I'm not fixing it here

https://github.com/apache/arrow/blob/master/cpp/src/parquet/util/comparison.cc#L68

Reply:
@fsaintjacques Can you please elaborate your concern here with the writer side? I authored this code originally.